File: Application/src/main/java/com/shareyourproxy/api/domain/factory/RealmGroupFactory.java
package com.shareyourproxy.api.domain.factory;

import com.shareyourproxy.api.domain.model.Group;
import com.shareyourproxy.api.domain.realm.RealmGroup;
import com.shareyourproxy.api.domain.realm.RealmString;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

import io.realm.RealmList;

/**
 * Factory for creating {@link RealmGroup}s.
 */
public class RealmGroupFactory {

    /**
     * Return a RealmList of Contacts from a user
     *
     * @param groupHashMap to get contacts from
     * @return RealmList of Contacts
     */
    public static RealmList<RealmGroup> getRealmGroups(HashMap<String, Group> groupHashMap) {
        if (groupHashMap == null) {
            return null;
        }
        RealmList<RealmGroup> realmGroupArray = new RealmList<>();
        for (Map.Entry<String, Group> entryGroup : groupHashMap.entrySet()) {
            Group group = entryGroup.getValue();
            RealmGroup realmGroup = new RealmGroup();
            realmGroup.setId(group.id());
            realmGroup.setLabel(group.label());
            realmGroup.setChannels(createRealmStringList(group.channels()));
            realmGroup.setContacts(createRealmStringList(group.contacts()));
            realmGroupArray.add(realmGroup);
        }
        return realmGroupArray;
    }

    private static RealmList<RealmString> createRealmStringList(HashSet<String> values) {
        if (values == null) {
            return null;
        }
        RealmList<RealmString> list = new RealmList<>();
        for (String id : values) {
            RealmString newChannel = new RealmString();
            newChannel.setValue(id);
            list.add(newChannel);
        }
        return list;
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/domain/factory/RealmGroupFactory.kt
package com.shareyourproxy.api.domain.factory

import com.shareyourproxy.api.domain.model.Group
import com.shareyourproxy.api.domain.realm.RealmGroup
import com.shareyourproxy.api.domain.realm.RealmString
import io.realm.RealmList
import java.util.*

/**
 * Factory for creating [RealmGroup]s.
 */
internal object RealmGroupFactory {

    /**
     * Return a RealmList of Contacts from a user
     * @param groupHashMap to get contacts from
     * @return RealmList of Contacts
     */
    fun getRealmGroups(groupHashMap: HashMap<String, Group>): RealmList<RealmGroup> {
        val realmGroupArray = RealmList<RealmGroup>()
        groupHashMap.entries.forEach {
            val group = it.value
            val realmGroup = RealmGroup()
            realmGroup.id = group.id
            realmGroup.label = group.label
            realmGroup.channels = createRealmStringList(group.channels)
            realmGroup.contacts = createRealmStringList(group.contacts)
            realmGroupArray.add(realmGroup)
        }
        return realmGroupArray
    }

    private fun createRealmStringList(values: HashSet<String>?): RealmList<RealmString> {
        val list = RealmList<RealmString>()
        return when (values) {
            null -> list
            else -> {
                for (id in values) {
                    val newChannel = RealmString()
                    newChannel.value = id
                    list.add(newChannel)
                }
                list
            }
        }
    }
}
